emacsPackages: don't set `preBuild` or `patches` to null

marienz opened the pull request:
Under some circumstances, some MELPA packages had their `preBuild` or `patches` attributes set to `null`. This does not seem to break their build, but it makes using `overrideAttrs` less convenient: the usual pattern of appending to `previousAttrs.preBuild or ""` does not work if `preBuild` is set but null.

This is the minimal fix: default these to `""` or `[]` instead. While I'm there, fix what looks like a copy/paste error in `bpr` (which will set `patches` to the `previousAttrs` attrset under some circumstances).

Instead of `if`/`else`, I think we could probably use `optionalAttrs`, along the lines of:

```nix
  dap-mode = super.dap-mode.overrideAttrs (
    finalAttrs: previousAttrs:
      lib.optionalAttrs (lib.versionOlder finalAttrs.version "20250131.1624") {
        preBuild = ''
          rm --verbose dapui.el
        ''
        + previousAttrs.preBuild or "";
    }
  );
```

I haven't tried that approach: please let me know if that looks better and I'll confirm it works and add a commit.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

Built the modified packages using

```
nix-repl> :b legacyPackages.x86_64-linux.emacs.pkgs.withPackages (epkgs: [epkgs.alectryon epkgs.codesearch epkgs.dap-mode epkgs.frontside-javascript epkgs.kanagawa-themes epkgs.org-noter epkgs.org-projectile epkgs.bpr ])
```

and confirmed the resulting Emacs started, but I don't know how to meaningfully test the Emacs packages themselves, sorry...

I *think* I applied everything in CONTRIBUTING.md that is relevant but might've missed something, I'm new here :)

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

marienz committed:
emacsPackages: don't set attributes to null

Depending on the version of the package from MELPA, nixpkgs can set
attributes like `preBuild` or `patches` to `null`. `mkDerivation`
filters out `null`, so this has the desirable effect of not
unnecessarily changing the actual derivation. But it is inconvenient
when using `overrideAttrs`: it breaks the common pattern of
`(previousAttrs.preBuild or "") + ''...''` (also used elsewhere in the
files this commit changes), which will fail to coerce `null` to a string
or list to append to.

Fix this everywhere attributes are set to null, by defaulting to `""` or
`[ ]` instead.

We cannot set the attribute conditionally, because the attributes names
returned by `overrideAttrs` cannot depend on `finalAttrs` (only the
values can).

This changes some helpers that are (indirectly) used by several
packages, but that part of the change is effectively a noop, because the
attributes involved are always set (either by the helper or on the
original derivation).

marienz committed:
emacsPackages.bpr: fix future preBuild

We conditionally set `preBuild` depending on version. For a new enough
version of the `bpr` package, we would set `preBuild` to the
`previousAttrs` attrset.

This looks like a copy/paste error, so fix it.

This code is currently not reachable: it would need a newer version of
`bpr`, which has had no changes upstream for several years.

acid-bong commented:
title conventions apply to both commits and PRs, you must rename your PR accordingly (it's absolutely not clear from the first glance what it's about)

marienz commented:
> title conventions apply to both commits and PRs, you must rename your PR accordingly

Fixed, my mistake, sorry about that. Shouldn't happen again.

(I clicked "create PR" before noticing the title conventions section in CONTRIBUTING.md, rewrote the commit messages accordingly, but missed the PR title based on an original commit message.)

panchoh commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 452898`
Commit: `778d1f55004c50faca751aa5af23f4283bc3f544`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>tests.pkgs-lib</li>
  </ul>
</details>

panchoh commented:
Hi, @marienz, good to see you here!

I understand this came across in the context of [marienz/nix-doom-emacs-unstraightened](marienz/nix-doom-emacs-unstraightened), where you have to custom-build many packages to fit Doom’s pins, right? (extremely grateful user here, thank you!).

I’m fine with the PR as it is, although I think that in the long run we would be better off with the approach you suggested involving `lib.optionalString` and `lib.optionalAttrs`, since they afford a nice reusable pattern, that could be easily adopted for new additions to this file, and hence help keep it consistent over time.

Nevertheless, I’d like to ask @jian-lin to share his thoughts on this, especially since he authored the affected lines.

panchoh reviewed (approved):
And good catch (`pbr`). Yes, seems like an oversight to me.

marienz commented:
> I understand this came across in the context of [marienz/nix-doom-emacs-unstraightened](https://github.com/NixOS/nixpkgs/pull/marienz/nix-doom-emacs-unstraightened), where you have to custom-build many packages to fit Doom’s pins, right?

Right. I didn't want to call that out as an example because the rest of what I'm doing there is not exactly... reasonable (nixpkgs shouldn't have to support it). But as far as I know this specific thing (assuming that if attributes like `preBuild` or `patches` are present, they are of the right type to append to) is pretty common (it's used elsewhere in the same file I'm changing in this PR). So it seemed worth changing on this side.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/melpa-packages.nix:
`lib-override-helper.nix` and `elpa-common-overrides.nix` also have this `or null` pattern.  Could you also fix them?

marienz replied:
The elpa-common-overrides.nix one looked straightforward enough I've squashed it into the existing commit (single affected package, same attributes). (I don't know nixpkgs preferences around squashing, please let me know if I'm squashing too much or not enough).

lib-override-helper.nix is a little more interesting: they're in helpers so this potentially changes more packages (and they're different attributes). But it looks like they're not *actually* changing evaluation: see reasoning in the new commit.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/melpa-packages.nix:
> I haven't tried that approach: please let me know if that looks better and I'll confirm it works and add a commit.

IIRC, your proposed approach causes infinite recursion.

I do not have preference for these two styles.  However, in some cases (say we need to apply different overrides for more than two versions), the proposed approach does not work.

marienz replied:
> IIRC, my motivation for or null pattern is to avoid unnecessary rebuilds. In other words, the .drv file does not change if we do not need to apply an override.

Ahh, I didn't realize that: I assumed that the `null` was actually ending up in `.attrs.json` and/or setting an environment variable to the empty string (that's what https://nix.dev/manual/nix/2.18/language/derivations#derivations says happens...), but I just tested and that's not the case (`.overrideAttrs { preBuild = null; }` does indeed not change the store hash, while `preBuild = ""` does).

[This](https://github.com/NixOS/nixpkgs/blob/ea4f36e630850c5946eb4ff1a50d9a734b6c2392/lib/derivations.nix#L175-L178) explains it (and that function does essentially what your original code was doing).

> your proposed approach causes infinite recursion.

...well, that's what I get for not testing it (whoops):

```
nix-repl> legacyPackages.x86_64-linux.emacsPackages.dap-mode
error:
       error: infinite recursion encountered
       at /nix/store/ls0g2dbm2bbsz2d0zkgk6vby1ih42pjy-source/pkgs/stdenv/generic/make-derivation.nix:84:22:
           83|       # An infinite recursion here can be caused by having the attribute names of expression `e` in `.overrideAttrs(finalAttrs: previousAttrs: e)
` depend on `finalAttrs`. Only the attribute values of `e` can depend on `finalAttrs`.
           84|       args = rattrs (args // { inherit finalPackage overrideAttrs; });
             |                      ^
           85|       #              ^^^^
```

That's unfortunate: if we do the version check against `previousAttrs.version` it works, but that seems incorrect... (if someone uses `.overrideAttrs` to build a different version we want this check to respect that).

So either we unconditionally set these to strings/lists (causing some rebuilds) or we set them to null (better for nixpkgs itself but inconvenient for `overrideAttrs` outside of it).

I've updated the commit messages to clarify what the original code was doing.

marienz replied:
Yeah, I figured that out while I was writing that comment (and didn't edit all of my initial confusion back out of the comment...)

jian-lin reviewed (changes requested):
Thanks for the fix!

IIRC, my motivation for `or null` pattern is to avoid unnecessary rebuilds.  In other words, the `.drv` file does not change if we do not need to apply an override.

It turns out the `or null` pattern is wrong.

jian-lin reviewed (approved):
> I don't think the change to lib-override-helper.nix will cause many rebuilds, but I may have missed something. I assume nixpkgs-ci (or some other bot) will automatically update the rebuild-* labels?

Nixpkgs CI only counts rebuilds for packages built by hydra.nixos.org.  Currently, we do not build elisp packages there so our CI will report 0 rebuilds.

I guess other tools, such as @panchoh's `nixpkgs-review` [report][1], have similar behavior.

This PR will causes some rebuilds.  It is not a big deal since we are fixing an issue here.

[1]: https://github.com/NixOS/nixpkgs/pull/452898#issuecomment-3417823368

jian-lin reviewed (approved):
Nice work! Thanks!
